lib/master/master.go

```diff
@@ -0,0 +1,460 @@
+}
+
+func makeHeartbeatHandler(c chan lib.Heartbeat) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var beat lib.Heartbeat
+		// TODO parse json and sends the results directly to the main loop
+		c <- beat
+	}
+}
+
+func makeStatusHandler(c chan chan statusSummary) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		resultsChan := make(chan statusSummary)
+		c <- resultsChan
+		<-resultsChan
+		// TODO read the results and serve status page
+	}
+}
+
+func (m *Master) updateOnHeartbeat(beat lib.Heartbeat) {
+	// check whether slave already exist, if not create one
+	// create a goroutine for every slave that checks for missed hearbeats
+	// if a miss is detected then report to master to be handled
+	// update task statuses on every heartbeat
+}
+
+func instanceIds(instances []*ec2.Instance) []*string {
+	instanceIds := make([]*string, len(instances))
+	for i, instance := range instances {
+		instanceIds[i] = instance.InstanceId
+	}
+	return instanceIds
+}
+
+func (m *Master) getTaskToSchedule() int {
+	for idx, t := range m.newTasks {
+		if !m.jobs[t.JobID].reachedMaxTasks() {
+			return idx
+		}
+	}
+	return -1
+}
+
+func (m *Master) slaveAvailable() string {
+	minimumTasks := math.MaxInt64
+	var slaveIP string
+	for ip, i := range m.instances {
+		if assignedTasks := len(i.tasks); assignedTasks < minimumTasks && assignedTasks < i.maxSlots {
+			minimumTasks = assignedTasks
+			slaveIP = ip
+		}
+	}
+	return slaveIP
+}
+
+func (m *Master) countTotalSlots() int {
+	cnt := 0
+	for _, i := range m.instances {
+		cnt += i.maxSlots
+	}
+	return cnt
+}
+
+func (m *Master) maxSlots() int {
+	// TODO do we set the manually or it's a property of AWS?
+	return 20 * lib.MaxSlotsPerInstance
+}
+
+func (m *Master) countRequiredSlots() int {
+	cnt := len(m.scheduledTasks)
+	cnt += len(m.newTasks)
+	if cnt > m.maxSlots() {
+		return m.maxSlots()
+	}
+	return cnt
+}
+
+// runController runs one iteration
+func (m *Master) runController() {
+	err := float64(m.countRequiredSlots() - m.countTotalSlots())
+
+	dt := m.controller.dt.Seconds()
+	m.controller.integral = m.controller.integral + err*dt
+	derivative := (err - m.controller.prevErr) / dt
+	output := m.controller.kp*err +
+		m.controller.ki*m.controller.integral +
+		m.controller.kd*derivative
+	m.controller.prevErr = err
+
+	log.Printf("err: %v, output: %v\n", err, output)
+	m.adjustInstanceCount(int(output))
+}
+
+func (m *Master) adjustInstanceCount(n int) {
+	if n > 0 {
+		go func() {
+			_, err := createSlaves(m.svc, n)
+			if err != nil {
+				log.Println("Failed to create slaves", err)
+			} else {
+				log.Printf("%v instances created successfully\n", n)
+				// no need to report back to the master loop
+				// because it should start receiving heartbeat messages
+			}
+		}()
+	} else {
+		// negate n to represent the (positive) number of instances to kill
+		// scale by the number of
+		n := -n
+		n = n / lib.MaxSlotsPerInstance
+		if n < 0 {
```

Comment:
This case will never be `true`. Also it does not return if it ever would reach the next line.